planner/cascades/transformation_rules.go

```diff
@@ -1448,3 +1452,174 @@ func (r *TransformLimitToTableDual) OnTransform(old *memo.ExprIter) (newExprs []
+	join := joinExpr.ExprNode.(*plannercore.LogicalJoin)
+
+	var innerChildIdx int
+	switch join.JoinType {
+	case plannercore.LeftOuterJoin:
+		innerChildIdx = 1
+	case plannercore.RightOuterJoin:
+		innerChildIdx = 0
+	default:
+		return nil, false, false, nil
+	}
+	outerGroup := joinExpr.Children[1^innerChildIdx]
+	innerGroup := joinExpr.Children[innerChildIdx]
+
+	outerUniqueIDs := set.NewInt64Set()
+	for _, outerCol := range outerGroup.Prop.Schema.Columns {
+		outerUniqueIDs.Insert(outerCol.UniqueID)
+	}
+	// only when agg only use the columns from outer table can eliminate outer join
+	if !isColsAllFromOuterTable(r.getAggCols(agg), outerUniqueIDs) {
+		return nil, false, false, nil
+	}
+	// outer join elimination with duplicate agnostic aggregate functions
+	if r.isAggFuncAllDupAgnostic(agg) {
+		newAggExpr := r.doEliminateOuterJoin(agg, join, joinExpr.Schema(), outerGroup, innerChildIdx)
+		return []*memo.GroupExpr{newAggExpr}, true, false, nil
+	}
+	// outer join elimination without duplicate agnostic aggregate functions
+	innerJoinKeys := extractInnerJoinKeys(join, innerChildIdx)
+	contain, err := isInnerJoinKeysContainUniqueKey(innerGroup, innerJoinKeys)
+	if err != nil {
+		return nil, false, false, err
+	}
+	if contain {
+		newAggExpr := r.doEliminateOuterJoin(agg, join, joinExpr.Schema(), outerGroup, innerChildIdx)
+		return []*memo.GroupExpr{newAggExpr}, true, false, nil
+	}
+
+	return nil, false, false, nil
+}
+
+// It will transform `Aggregation->Join->outer,inner` to `Aggregation->outer` or `Aggregation->Selection->outer`
+func (r *EliminateOuterJoinBelowAggregation) doEliminateOuterJoin(agg *plannercore.LogicalAggregation, join *plannercore.LogicalJoin, joinSchema *expression.Schema, childGroup *memo.Group, innerChildIdx int) *memo.GroupExpr {
+	newAggExpr := memo.NewGroupExpr(agg)
+	ok, sel := convertOuterJoinToSel(join, innerChildIdx)
+	if ok {
+		selExpr := memo.NewGroupExpr(sel)
+		selExpr.SetChildren(childGroup)
+		selGroup := memo.NewGroupWithSchema(selExpr, joinSchema)
+		childGroup = selGroup
+	}
+	newAggExpr.SetChildren(childGroup)
+	return newAggExpr
+}
+
+// extract cols from agg
+func (r *EliminateOuterJoinBelowAggregation) getAggCols(agg *plannercore.LogicalAggregation) (aggCols []*expression.Column) {
+	for _, groupByItem := range agg.GroupByItems {
+		aggCols = append(aggCols, expression.ExtractColumns(groupByItem)...)
+	}
+	for _, aggDesc := range agg.AggFuncs {
+		for _, expr := range aggDesc.Args {
+			aggCols = append(aggCols, expression.ExtractColumns(expr)...)
+		}
+	}
+	return aggCols
+}
+
+// return true only all the aggregate functions are duplicate agnostic.
+// Only the following functions are considered to be duplicate agnostic:
+//   1. MAX(arg)
+//   2. MIN(arg)
+//   3. FIRST_ROW(arg)
+//   4. Other agg functions with DISTINCT flag, like SUM(DISTINCT arg)
+func (r *EliminateOuterJoinBelowAggregation) isAggFuncAllDupAgnostic(agg *plannercore.LogicalAggregation) bool {
```

Comment:
Can we reuse the function in `rule_outer_join_elimination.go`?

Reply:
function `getDupAgnosticAggCols` in `rule_join_elimination.go` must be called by  `outerJoinEliminator`.

In function `getDupAgnosticAggCols` , it checks whether a LogicalPlan is LogicalAggregation, all aggregate functions are duplicate agnostic and extracts all the columns from the duplicate agnostic aggregate functions.
Here, we just need checks all aggregate functions are duplicate agnostic.

Reply:
Yes, but we can still reuse it, right? All we should do is to ignore the extra check parts.